Use possibly cached index trees in watch-point

oguzziya opened the pull request:
## Main changes of this PR
- Change closest find functions to use already cached index trees
- Use same functionality as nearest-projection mapping for interpolation

## Motivation and additional information
For the calculation of the watch point, we use closest point/edge/triangle finding. For nearest-projection and nearest-neighbor mappings index trees are possibly built up therefore we can use them to efficiently get the corresponding primitives.

Additionally, nearest-projection mapping and watch-point calculation involves similar operations, both using interpolation to determine the value. Therefore, using consistent functions across the same operation helps to improve maintainability.

## Author's checklist

* [x] I added a changelog file with this PR number in `docs/changelog/` if there are noteworthy changes.
* [x] I ran `tools/formatting/check-format` and everything is formatted correctly.
* [x] I sticked to C++14 features.
* [x] I sticked to CMake version 3.10.
* [x] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

<!-- Tag people next to each point and add points for specific questions -->

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?
* [ ] (more questions/tasks)

fsimonis commented on src/precice/impl/WatchPoint.cpp:
```diff
@@ -61,13 +59,15 @@ void WatchPoint::initialize()
   _weights.clear();
   _vertices.clear();
 
+  mesh::Vertex watchVertex(_point, 0);
```
Having to create a dummy point is a strong indicator for the interface of `query::Index` being incomplete.
Looking at it, there is really no reason why the `Index::getClosestX` functions require a full Vertex as argument (aside from convenience).

My suggestion for now:
* We add overloads that take `const Eigen::VectorXd&` instead of `const mesh::Vertex&`.
* We can then implement the Vertex version then using the Eigen version.

fsimonis commented on src/precice/impl/WatchPoint.cpp:
```diff
@@ -99,33 +99,33 @@ void WatchPoint::initialize()
-  if (findEdge(*_mesh)) {
-    if (findEdge.getEuclidianDistance() < _shortestDistance) {
-      //         _closestEdge = & findEdge.getClosestEdge ();
+  auto closestEdges = indexTree.getClosestEdges(watchVertex, 1);
```
The edge/triangle with the shortest distance may not be an interpolating projection.
Have a look at the implementation in the nearest projection mapping.

oguzziya commented:
There are a few critical changes that need suggestion:
- For the time being, 3 new member functions are added to the `Index` class to handle the projection element computation: `findVertexProjectionElements()`, `findEdgeProjectionElements()`, `findTriangleProjectionElements()`. If edge cannot find appropriate projection element, fall backs to vertex and returns corresponding vertex match and interpolation element. Similarly, if an appropriate triangle projection element cannot be found, the first fall backs to the edge, then the vertex if necessary.
- This brings the problem of return type since we identify each match type with different tags. For the time being, I added the `GenericMatch` tag in order to handle this. This part is open to suggestions.
- Although having these functions in the `Index` class is extremely handy, I think this is not the best place.

fsimonis commented on src/query/Index.cpp:
```diff
@@ -91,14 +75,84 @@ std::vector<TriangleMatch> Index::getClosestTriangles(const mesh::Vertex &source
   return matches;
 }
 
+VertexMatch Index::getClosestVertex(const mesh::Vertex &sourceVertex)
```
Please implement using the Eigen version.

fsimonis commented on src/query/Index.cpp:
```diff
@@ -91,14 +75,84 @@ std::vector<TriangleMatch> Index::getClosestTriangles(const mesh::Vertex &source
+  }
+}
+
+VertexMatch Index::getClosestVertex(const Eigen::VectorXd &sourceCoord)
```
There is a more straight-forward implementation for this exact case, as the nearest neighbour is always the nearest projection for vertices.

Have a look at the old implementation:

https://github.com/precice/precice/blob/6fdeb97c20a92ecc47f55bdb00a8f95cc60b77af/src/mapping/NearestProjectionMapping.cpp#L141-L146

fsimonis commented on src/query/Index.cpp:
```diff
@@ -91,14 +75,84 @@ std::vector<TriangleMatch> Index::getClosestTriangles(const mesh::Vertex &source
+    return matches.back();
+  } else {
+    // return a default match with empty index
+    return VertexMatch(-1, std::numeric_limits<double>::max());
```
Let's add an assertion here. The caller should check this.

fsimonis commented on src/query/Index.cpp:
```diff
@@ -91,14 +75,84 @@ std::vector<TriangleMatch> Index::getClosestTriangles(const mesh::Vertex &source
+  }
+}
+
+EdgeMatch Index::getClosestEdge(const Eigen::VectorXd &sourceCoord)
```
The singular version is not required. [YAGNI](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it)

fsimonis commented on src/query/Index.cpp:
```diff
@@ -91,14 +75,84 @@ std::vector<TriangleMatch> Index::getClosestTriangles(const mesh::Vertex &source
+  }
+}
+
+TriangleMatch Index::getClosestTriangle(const Eigen::VectorXd &sourceCoord)
```
Also not needed

fsimonis commented on src/query/Index.hpp:
```diff
@@ -40,23 +41,32 @@ class Index {
   std::vector<size_t> getVerticesInsideBox(const mesh::BoundingBox &bb);
 
+  /// Temporary before other refactor, return Interpolation elements and the distance
+  std::pair<InterpolationElements, GenericMatch> findVertexProjectionElements(const mesh::Vertex &sourceVertex, int n);
+  std::pair<InterpolationElements, GenericMatch> findEdgeProjectionElements(const mesh::Vertex &sourceVertex, int n);
+  std::pair<InterpolationElements, GenericMatch> findTriangleProjectionElements(const mesh::Vertex &sourceVertex, int n);
```
You are right regarding the `GenericMatch` it really is pointless.
Lets use only the distance to the matched primitive `std::pair<InterpolationElements, double>` for now. If we ever need the type of the matched primitive, we should add an enum to the GenericMatch which identifies the primitive it matched against.

fsimonis reviewed (changes requested):
This looks really good now!
Just a few changes and we are ready to merge.

oguzziya commented:
I made the following changes:
- Removed unnecessary closest finding functions. Actually, with the refactored projection element finding functions, explicit calls on closest finding functions are not necessary and can be implemented as private functions. However, it will cause some changes in tests, therefore I plan it to do in another PR.
- I take one step back from the projection element finding functions. Previously, falling back to one lower dimension was done inside functions. However, this behaviour is a little bit hideous and might cause errors in the future. In addition, watch point follows different approach, and not finding edge or triangle projection elements would cause one unnecessary calculation of vertex projection element. Therefore, falling back to lower dimensions is done in the neares projection mapping. 
- In watchpoint, vertices and weights were stored seperately. However, with the implementation of the interpolations elements, it is not necessary and it is refactored accordingly.

fsimonis commented:
> I take one step back from the projection element finding functions. Previously, falling back to one lower dimension was done inside functions. However, this behavior is a little bit hideous and might cause errors in the future.

The issue here is that this implementation requires a sentinel value, which requires to be checked by the caller, which in turn is dangerous. The resulting code is also less readable in both the function and the call-site.
The nearest-projection has the nice property that there is always a nearest projection, so lets use it.
This will get a bit more complicated when we eventually implement projection to volumes #468.

Anyhow, a good function name and documentation is the solution here.

> In addition, watchpoint follows a different approach, and not finding an edge or triangle projection elements would cause one unnecessary calculation of vertex projection element. Therefore, falling back to lower dimensions is done in the nearest projection mapping. If you do not like it I can switch it back to the implicit one.

Let's not worry about the runtime of the watchpoint initialization. Its implementation could be simplified anyhow (Find nearest projection, determine owner, done), but that is for another time.

oguzziya commented:
- I changed the function name that is going to be used in the library to `findNearestProjection()`. The individual functions are still in the interface since otherwise, the implementation gets dirty and there is code duplication. In addition, in the case of future volume coupling, it is easier to extend the functions without having to add so many if conditions.
- With this change, I also simplified the `WatchPoint` initialization, we directly find the projection element, then the owner.

fsimonis commented on src/precice/impl/WatchPoint.cpp:
```diff
@@ -93,35 +88,6 @@ void WatchPoint::initialize()
   }
 
   PRECICE_DEBUG("Rank: " << utils::MasterSlave::getRank() << ", isClosest: " << _isClosest);
```
We could clear the `_interpolationElements` when `isClosest == false` this may prevent some mistakes in the future.

fsimonis commented on src/precice/impl/WatchPoint.cpp:
```diff
@@ -131,12 +97,12 @@ void WatchPoint::exportPointData(
   Eigen::VectorXd coords = Eigen::VectorXd::Constant(_mesh->getDimensions(), 0.0);
-  for (size_t i = 0; i < _vertices.size(); i++) {
-    coords += _weights[i].weight * _vertices[i]->getCoords();
+  for (const auto elem : _interpolationElements) {
```
```suggestion
  for (const auto& elem : _interpolationElements) {
```
No need to copy every element. Same below.

fsimonis reviewed (approved):
This refactoring really improves readability by __a lot__.
Really nicely done!
